library/core/src/ptr/mod.rs

```diff
@@ -63,12 +63,212 @@
+//! casts from an address to a pointer are essentially just applying offsets/indexing.
+//!
+//! This is generally trivial to do for simple cases like tagged pointers *as long as you
+//! represent the tagged pointer as an actual pointer and not a usize*. For instance:
+//!
+//! ```
+//! #![feature(strict_provenance)]
+//!
+//! unsafe {
+//!     // A flag we want to pack into our pointer
+//!     static HAS_DATA: usize = 0x1;
+//!     static FLAG_MASK: usize = !HAS_DATA;
+//!
+//!     // Our value, which must have enough alignment to have spare least-significant-bits.
+//!     let my_precious_data: u32 = 17;
+//!     assert!(core::mem::align_of::<u32>() > 1);
+//!
+//!     // Create a tagged pointer
+//!     let ptr = &my_precious_data as *const u32;
+//!     let tagged = ptr.map_addr(|addr| addr | HAS_DATA);
+//!
+//!     // Check the flag:
+//!     if tagged.addr() & HAS_DATA != 0 {
+//!         // Untag and read the pointer
+//!         let data = *tagged.map_addr(|addr| addr & FLAG_MASK);
+//!         assert_eq!(data, 17);
+//!     } else {
+//!         unreachable!()
+//!     }
+//! }
+//! ```
+//!
+//! Something more complicated and just generally *evil* like a XOR-List requires more significant
+//! changes like allocating all nodes in a pre-allocated Vec or Arena and using a pointer
+//! to the whole allocation to reconstitute the XORed addresses.
+//!
+//! Situations where a valid pointer *must* be created from just an address, such as baremetal code
+//! accessing a memory-mapped interface at a fixed address, are an open question on how to support.
+//! These situations *will* still be allowed, but we might require some kind of "I know what I'm
+//! doing" annotation to explain the situation to the compiler. Because those situations require
+//! `volatile` accesses anyway, it should be possible to carve out exceptions for them.
+//!
+//! Under [Strict Provenance] is is Undefined Behaviour to:
+//!
+//! * Access memory through a pointer that does not have provenance over that memory.
+//!
+//! * [`offset`] a pointer to an address it doesn't have provenance over.
+//!   This means it's always UB to offset a pointer derived from something deallocated,
+//!   even if the offset is 0. Note that a pointer "one past the end" of its provenance
+//!   is not actually outside its provenance, it just has 0 bytes it can load/store.
+//!
+//! But it *is* still sound to:
+//!
+//! * Create an invalid pointer from just an address (see [`ptr::invalid`][]). This can
+//!   be used for sentinel values like `null` *or* to represent a tagged pointer that will
+//!   never be dereferencable.
+//!
+//! * [`wrapping_offset`][] a pointer outside its provenance. This includes invalid pointers
+//!   which have no provenance. Unfortunately there may be practical limits on this for a
+//!   particular platform ([CHERI][] may mark your pointers as invalid, but it has a pretty
+//!   generour buffer that is *at least* 1KB on each side of the provenance). Note that
```

Comment:
Just a pedantic note: CHERI extending 32-bit RISC-V (with 64-bit capabilities) has fewer bits for bounds so it's not quite 1k there, but I don't think that really matters for this documentation :)

Reply:
Mmm. Having seen the kind of arguments that break out because of precisely the documentation, I actually think it would in fact be useful to be precise on this matter, if you can help us there.

Reply:
What I meant to say is that the actual limit is extremely architecture-dependent (and also depends on the alignment of your pointer), so the Rust documentation does not need to be the source of truth for that.
For example Arm's Morello architecture allows for at least 4K (or possible 8, I'd have to double-check) since they use more bits for bounds compared to CHERI-RISC-V.

If you want to be precise you could use `2^(floor(bounds_bits/2)−3)` as the absolute minimum, but in practice that will be higher, since aligned pointers can go further out-of bounds.

C/C++ only allow one-past-the-end pointers, but CHERI is a bit more lenient there. Maybe something like this?
```suggestion
//!   generous buffer on each side of the pointer bounds -- the exact range depends on
//!   the alignment of the pointer as well as the underlying architecture). Note that
```